### PR TITLE
Swift support

### PR DESCRIPTION
**Pull request by iMostfa**
Hi, 
i liked this plugin so much, so i tried to configure it for Swift. 

i ended up producing the query in my PR. 

but i have some questions as it's my first time to deal with the query stuff: 
- does the code look ok ? 
- i think that i'm using a lot of wild cards, so it works, but is it slow. how can i profile it ? 

Screenshot 2024-10-23 at 02 46 53

### CODE REVIEW

**Comment by HiPhish** [comment]
Thank you. Can you please add some sample Swift file which contains all of of the patterns you have defined? It does not have to make sense, but ideally it should be valid Swift that would compile; this is not a hard requirement but it would be nice so that language servers and linters don't mark things as errors.

> i think that i'm using a lot of wild cards, so it works, but is it slow.

It should work without the wildcards

> how can i profile it ?

Honestly, I have never had the need to profile the queries. The only time parsing got noticeable slow was Zig, but that was a problem with the parser. After the parser got an updated the performance in Zig became normal.

File: queries/swift/rainbow-delimiters.scm [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  "{" @delimiter
+  (_)*
+  "}" @delimiter @sentinel) @container
+
+;; Handle nested enums within classes or other enums
+(class_declaration
+  (enum_class_body
+    "{" @delimiter
+    (_)*
+    "}" @delimiter @sentinel)) @container
```
Comment (HiPhish) [inline queries/swift/rainbow-delimiters.scm:79]: Again, I do not understand why you need this nesting.

File: queries/swift/rainbow-delimiters.scm [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+(enum_entry
+  "{" @delimiter
+  (_)*
+  "}" @delimiter @sentinel) @container
+  
+;; Handle nested enum entries with their respective bodies
+(enum_entry
+  "{" @delimiter
+  (_)*
+  "}" @delimiter @sentinel) @container
```
Comment (HiPhish) [inline queries/swift/rainbow-delimiters.scm:91]: Duplicates. Also, I cannot find any `enum_entry` with curly braces in your sample.

File: queries/swift/rainbow-delimiters.scm [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+;; Support for didSet and willSet blocks
+(willset_didset_block
+  "{" @delimiter
+  (_)*
+  "}" @delimiter @sentinel) @container
+
+(didset_clause
+  "{" @delimiter
+  (_)*
+  "}" @delimiter @sentinel) @container
```
Comment (HiPhish) [inline queries/swift/rainbow-delimiters.scm:102]: There are not `willset_didset_block` and `didset_clause` nodes in your sample file.

**Review by HiPhish — COMMENTED** [review]
A couple of general notes:

- You don't need the `(_)*` wildcards
- When you have a pattern like `(foo "{" @delimiter)` the brace must be a child of `foo`, not a general descendant
- A pattern like `(bar "{")` is preferable over `(foo (bar "{"))` because it will match more; use the latter only if you really mean "only a `bar` that is a child of a `foo` and nothing else"

File: queries/swift/rainbow-delimiters.scm [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+;; Handle deeply nested functions
+(function_declaration
+  (function_body
+    "{" @delimiter
+    (_)*
+    "}" @delimiter @sentinel)) @container
+
```
[the remark this answers is not in the document]
Comment (iMostfa) [reply queries/swift/rainbow-delimiters.scm]: correct, removed

File: queries/swift/rainbow-delimiters.scm [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  (lambda_literal
+    "{" @delimiter
+    (_)*
+    "}" @delimiter @sentinel)) @container
+
+(call_suffix
+  (lambda_literal
+    "{" @delimiter
+    (_)*
+    "}" @delimiter @sentinel)) @container
```
[the remark this answers is not in the document]
Comment (iMostfa) [reply queries/swift/rainbow-delimiters.scm]: correct, as i said, i initially brute-forced it

File: queries/swift/rainbow-delimiters.scm [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  "[" @delimiter
+  (_)*
+  ":" @delimiter
+  (_)*
+  "]" @delimiter @sentinel) @container
+
+(tuple_type
+  "(" @delimiter
+  (_)*
+  ")" @delimiter @sentinel) @container
```
[the remark this answers is not in the document]
Comment (iMostfa) [reply queries/swift/rainbow-delimiters.scm]: added array_type, and dict_tpye, along with the example: 
Screenshot 2024-11-06 at 23 31 18

**Comment by iMostfa** [comment]
Thank you @HiPhish  for the review, i was already working on updating and fixing the problems i found after writing the example file, but forgot to put it on Draft xD

i applied almost all comments, added support For Arrays, and dict

this's how it looks now. 
image

i still missing the tuples matching, i'm trying to get it working as expected.

**Comment by HiPhish** [comment]
There were still a few mistakes and missing patterns. Please verify the commits I have pushed. If everything is in order rebase on the current `master` and squash the commits into one. I can merge the PR then.

**Comment by iMostfa** [comment]
i checked it, it looks good to me !
I think it's ready for production <3

**Comment by HiPhish** [comment]
Merged, thank you.

**Pull request merged**